### ISSUE TITLE
Consider moving events out of etcd

### ISSUE THREAD

**Issue by bgrant0607** [kubernetes/kubernetes#4432]
We've discussed this in person a few times.

Events account for the overwhelming majority of our etcd write volume, which is causing a variety of performance and stability problems (#3853, #3329, #2298, #3521, #4073). 

Since we GC events, they can't be relied upon as a messaging mechanism. Distributed control components should be watching status instead. Also, we support hooks to facilitate user extensions, so events are (arguably) not needed for that case, either, provided we add the necessary hooks. 

Therefore, the main purpose of events is to inform users of relevant occurrences. Given that, we could consider a different delivery mechanism for the events.

One proposal is to treat them as log messages. That has a certain appeal. We do "log" all such events into our equivalent of BigQuery internally, to facilitate searching, filtering, sorting, etc. 

We'd need to find a way to make the logs/events accessible via the CLI and UI, and we'd need a non-BigQuery implementation for other platforms. We could probably use fluent as the plugin mechanism. There's also an [amqp fluent plugin](https://github.com/restorando/fluent-plugin-amqp), which would facilitate pushing the events into a message queue.

@lavalamp @brendandburns @saad-ali @smarterclayton

**Comment by saad-ali** [comment]
cc @dchen1107

**Comment by lavalamp** [comment]
To solve the short-term problem, we could just make a specific etcd cluster for events. We've talked about sharding the etcd key space like this before.

**Comment by derekwaynecarr** [comment]
@smarterclayton and I have discussed sharding on key type as well as a future solution to this type of issue.

**Comment by derekwaynecarr** [comment]
I guess what I am curious on is if sharding events into a separate etcd cluster helps minimize etcd memory pressures, or is it just minimizing the issues around the etcd watch window?  I guess what I am wondering is if there is something unique about the write rate of events that makes etcd a bad fit.  

And if that is the case, do we see any other resources with similar write rates?

**Comment by dchen1107** [comment]
I agreed that etcd causes a lot of issues to our stability, and agreed that events generates a lot of volume being written to etcd. I also agreed that sharding on key type might be the right approach to go. 

But moving events out of etcd doesn't really solve our performance and reliability issues. See #4188 my last comment. I built a cluster last Friday, and scheduled 300 pods to 4 nodes. After a weekend, I came back and found tons of etcd issues. Even there is no new events generated at all, and existing events are deleted from etcd by hitting 2-days TTL. In this case, there are not much writes at all, but a lot of reads, I still hit etcd issue which causes podStatus flaps at master side. Anyway, I sent a list of requirements / questions to coreos side, hopefully can get engineer support.

**Comment by derekwaynecarr** [comment]
@dchen1107 - I am curious to know what you find in this area.

/cc @danmcp

**Comment by smarterclayton** [comment]
I definitely would like to help resolve stability issues.  I think the watch window should really be solved upstream, and that we could (at worst) design a fix for it.

----- Original Message -----

**Comment by dchen1107** [comment]
@derekwaynecarr and @smarterclayton will definitely sync-up with you on the status.

cc/ @lavalamp

**Comment by smarterclayton** [comment]
Also keep in mind that as soon as we switch to writing pod status to etcd, we are going to increase the pod write load by a factor of 5 or more.  So just moving events out won't solve the problem - events just exposes it first.

**Comment by xiang90** [comment]
@bgrant0607 I would like to help on this. But I need to know a rough idea of the current/future use case of etcd. We would like to take time to discuss and help to improve in the following months. 

@smarterclayton I can do a quick "improvement" for watching index advance notification. But after looking at the etcd helper function in k8s and how kubelet is using it, I think it might not be a big issue? Moreover, increasing the window size will not help. It will just increase the memory usage.

I do look around the codebase. I would appreciate if someone would point me to some related files.

**Comment by bgrant0607** [comment]
@smarterclayton It's not just the write load of events that is causing the trouble, but the unbounded growth in number of objects -- hence all the work on redundancy elimination, eviction, etc. Status writes don't have those problems.

**Comment by bgrant0607** [comment]
@xiang90 Thanks for the offer to help. @dchen1107 and @saad-ali are leading the short-term work on etcd scaling issues on our side. If it would help, we could also provide an overview explanation of our usage and/or an overview of the longer-term direction.

**Comment by smarterclayton** [comment]
Aren't we using ttls?  Doesn't this just mean the TTL is too long if it's unbounded?

----- Original Message -----

**Comment by quinton-hoole-google** [comment]
Do we have a good handle yet on what the empirical scaling bounds of etcd are, and what the implied scaling bounds of kubernetes are?  If these are dumb questions, my only defence is that "I'm new around here" :-)

I'm thinking along the lines of a basic load/scaling test on etcd, plotting memory consumption, read and write latency etc against number of etcd nodes, watchers etc.  And then tying this back to what kubernetes cluster size, rate of change, API load etc this translates to.

Does such a study/document exist yet?

**Comment by quinton-hoole-google** [comment]
In case it's useful, here's an example of a similar study done on zookeeper many years ago:
http://wiki.apache.org/hadoop/ZooKeeper/ServiceLatencyOverview
I quick search around the interwebs did not reveal anything similar for etcd, but maybe someone else has  some pointers.

**Comment by smarterclayton** [comment]
Yes, Derek did this against 0.4.6 and we have a team at red hat that is doing some work.  The roadmap identifies the expected scale, but I don't know if we've done a recent work up that translates normal Kube operation down to a read write ratio and the expected characteristics.  Off the cuff the current event work is generating tens of events (writes) per minute, while node status and pod status would drive 2-5 writes per pod creation (on top of the initial create and the two writes the scheduler does).  The controllers generate tens of bulk reads every few minutes of a recursive subtree containing one particular type (where N might be thousands at the 1.0 target scale of hundreds of nodes with tens of pods).  Client requests typically generate one to three reads, but that workload is probably insignificant compared to the total workload.

**Comment by bgrant0607** [comment]
@smarterclayton Yes, the TTLs are/were too long.

**Comment by bgrant0607** [comment]
But making them a lot shorter without doing anything else dramatically reduced the utility of events.

**Comment by smarterclayton** [comment]
Right.

----- Original Message -----

**Comment by pmorie** [comment]
Event compression will also help to mitigate uncontrolled growth.

**Comment by derekwaynecarr** [comment]
I collected information on etcd 0.4.6 and etcd 0.5.0 late last year largely focused on memory consumption as it relates to the number of nodes since etcd held everything in memory.  

Given an etcd snapshot on disk of the entire data set, it appeared to consume 13x more memory at best when loaded in memory.  There was some variation depending on number of nodes in a directory and number of directories. 

Since every node is holding a string representation of our data, it grows very quickly.  I know etcd is looking to support not holding everything in memory in the new model so that should improve longer term.  I am excited to try that out in future.  

I tested Kube style scenarios with large numbers of nodes in etcd that most likely exceed our 1.0 goals.  200k to 1mill nodes across multiple directories so it would be good to get new data on 1.0 targets which we are doing.  

Other things of note were depending on size of data and the snapshot interval, you see large spikes in memory usage probably related to garbage collection in Go at each snapshot interval.  That was go 1.3.

**Comment by davidopp** [comment]
If we had some kind of archival tool that sucked the events into longer-term storage then we could set the TTLs lower. We have something like this internally that's pretty simple but effective (it operates on the whole transaction log, but same idea). Of course there is a lower bound on a reasonable TTL; if even a TTL like 1 hour overloads etcd, then futzing with TTL is not the right approach.

As @pmorie (indirectly) asked, have we re-evaluated all of this with #4073 in?

BTW the 13x memory blowup is nothing short of astounding (unless etcd is doing some kind of super-aggressive compression that makes the on-disk representation abnormally succinct). We'd probably be helping a lot of people if we had someone look into trying to reduce that.

**Comment by smarterclayton** [comment]
----- Original Message -----
> If we had some kind of archival tool that sucked the events into longer-term
> storage then we could set the TTLs lower. We have something like this
> internally that's pretty simple but effective (it operates on the whole
> transaction log, but same idea). Of course there is a lower bound on a
> reasonable TTL; if even a TTL like 1 hour overloads etcd, then futzing with
> TTL is not the right approach.

I think our intention over time is to suck both events and the full transaction log into two distinct stores - the former into one designed for high volume search, and the latter into a structured one that allows historical retrieval of older versions.  When etcd acquires the ability to control which versions remain in the cluster (for retrieval) we would probably leave the latter in etcd.  In the short term, we have to have a solution.  I don't think they have to be different, but I do think high volume events and low volume transaction record are distinct use cases.

> 
> As @pmorie (indirectly) asked, have we re-evaluated all of this with #4073
> in?
> 
> BTW the 13x memory blowup is nothing short of astounding (unless etcd is
> doing some kind of super-aggressive compression that makes the on-disk
> representation abnormally succinct). We'd probably be helping a lot of
> people if we had someone look into trying to reduce that.

Yes - I think it behooves us to assist the etcd team to identify and target issues, and also to help with the forthcoming store work.  Most practical clusters are going to be <1million keys, and based on discussions so far that's achievable with the upstream store proposal.

**Comment by dchen1107** [comment]
Here is the current proposal I had for going forward on investigation: 

1) Introduce flag to control whether writing events to etcd or not; by default, no.
2) Events by default will be logged to disk, fluentd-elasticsearch will collect such logs. The user / client tools can collect the events related pods, nodes, and other objects through queries.
3) For users without fluentd-elasticsearch as addon service, you still can config their cluster to get events through etcd for now. 

cc/ @bgrant0607 @saad-ali @brendandburns

**Comment by lavalamp** [comment]
I don't know if I've written this down before, so I'll say it here. IMO, kubernetes should store events only long enough to enable some utility to put the events in 3rd party storage (accounting for downtime SLAs etc). kubernetes itself should NOT be responsible for event storage. Much like you must first list and then watch, to use events you must first observe the current state of the thing you care about and only then pay attention to events. 

I think we should:
* Set the event TTL low, like 30 minutes. (IMO, if you want to see events older than this you really need to have set up some infinite-event-storage system.)
* Add Dawn's drop-events-on-the-floor flag, but default off.
* Do an experiment to see if turning the flag off really improves system performance.

The advantage to this is that it's really easy, and it will let us focus on solving more pressing issues; more complete solutions to this can then come after 1.0.

**Comment by a-robinson** [comment]
Step 1 should definitely be doing a couple perf tests with and without events enabled to see their effect. I doubt that shortening the TTL could be a long-term solution, since all it helps with is memory pressure. It doesn't solve the write-rate or watch index problems.

**Comment by lavalamp** [comment]
> ... It doesn't solve the write-rate or watch index problems.

But these are problems we will have in larger clusters event *without* events, so it is good to find them now. Specifically, any problem with watch index is a *bug* in the code doing the watch. I am skeptical that we have problems with "write rate" and want to see some evidence. If we have those problems, we will have to fix them other ways anyway: writing object status to etcd will certainly trigger both watch index problems and "write-rate" problems.

To recap: the only problem unique to events in etcd is large memory size.

**Comment by lavalamp** [comment]
I also want to toss out something @smarterclayton mentioned in person yesterday, which is:

If etcd makes its entire log history available, then instead of writing a new event for every object, we can have one event per object and update the existing event for every event that comes in. This way all of the events are available in the etcd history (presumably stored on disk), but only the latest one takes up memory.

**Comment by smarterclayton** [comment]
+1 - let's root out the core issues, fix watch in etcd so it gives us a sequence update to refresh our watch window; and test.

We can also bulk write events and work with etcd guys to get a multi put (which is effectively what influx is doing) and we can reduce our writes by at least a large constant factor

**Comment by davidopp** [comment]
Does anyone on Kubernetes own this issue? As in, working on it full-time. If not, let's ask someone who is currently working on something that's not for 1.0. I have a couple of candidates in mind. We have been discussing it passionately for a while but it's lacking an owner to move it forward. I'm getting worried that we're going to make a bad decision because we're not willing to put in the time to first see how far we can go with etcd.

**Comment by xiang90** [comment]
> If etcd makes its entire log history available, then instead of writing a new event for every object, we can have one event per object and update the existing event for every event that comes in. This way all of the events are available in the etcd history (presumably stored on disk), but only the latest one takes up memory.

@lavalamp I talked with @bgrant0607 about this yesterday. 
The plan is to provide a user controllable compaction mechanism. So that etcd keeps all the operation history until a compaction.
However, we are still designing the v3 api/store.

> fix watch in etcd so it gives us a sequence update to refresh our watch window

The window slicing notification is another thing as far as I see.

**Comment by xiang90** [comment]
@davidopp 

> how far we can go with etcd.

You should have a rough idea about the etcd's status pre k8s 1.0.
etcd is still involving fast especially we just finished making its internal "solid". 

But there are something you should not put into etcd in general. For something that does not require sequential consistency or even read/write consistency and reliability, you probably want to move it out of etcd.

**Comment by lavalamp** [comment]
@xiang90 thanks for the clarification.

> The window slicing notification is another thing as far as I see.

I see this as a very nice to have optimization, but since all of our code should already list, it's not essential.

**Comment by a-robinson** [comment]
cc @kelseyhightower given our lunch conversation about this issue

**Comment by a-robinson** [comment]
@davidopp No one currently owns it, and someone definitely should. I'm very interested, but don't currently have the time.

**Comment by davidopp** [comment]
Saad, I see you assigned this to yourself. Do you have time to work on this full-time? BTW the "this" I'm referring to isn't just events, but getting a handle on the whole etcd stability and scalability issue, tuning the k8s and etcd code as needed, etc.

**Comment by saad-ali** [comment]
Yep, I'll take ownership of this.

Immediate plan of action is pretty much what you proposed:
* Output events to both etcd and log files gated by flags (so we can enable/disable either).
* Setup fluentd-elasticsearch to process the new events log.
* Run some tests to see the delta between 1) events output to etcd with current 2 day TTL, 2) no events output to etcd, and 3) events output to etcd with a 2 hr TTL.
  * Will monitor avg #read/writes, total memory utilization, etc.
  * If we find that even the reduced number of events cause issues, we will remove the ability to store events in etcd all together (to ensure etcd is stable for 1.0 release).
  * If we find that reducing the number of events is sufficient, to keep etcd running stable, we can enable both etcd and logs, where etcd is for quick access via kubectl to recent events, and older events can be found in logs.

**Comment by lavalamp** [comment]
Let's chat in person tomorrow.

**Comment by davidopp** [comment]
Yeah, let's discuss more -- I'd really like somebody to not just work on the event side of this but also on benchmarking etcd (the etcd folks were kind enough to give us some code, but I didn't want to link to it from Github since I wasn't sure if they wanted it publicly available), understanding the scaling limits of etcd in general and specifically for k8s's usage pattern (both with and without using etcd for events), understanding what's going on server-side when we see "All the given peers are not reachable (Tried to connect to each peer twice and failed" (it's probably multiple different things at different times, based on what the etcd guys said yesterday), etc. Possibly a couple of people should split up the work, since this seems to be a critical piece of at least some of the reliability issues we've seen and there may be a lot of work to do (both in better understanding the problem and then working with the etcd guys to address any issues that are found). Note that I'm not making any assumptions about whether the principal problem is on the k8s side or the etcd side, but we're totally in the dark right now and it seems there's a lot of work to do.

**Comment by dchen1107** [comment]
@saad-ali You also need to make sure kubectl describe can retrieve related events to such object; otherwise I think this is a regression for introspection. 

Also agreed with @davidopp. Saad, I just forwarded you an email about source code on benchmarking etcd. We can start from there.

**Comment by zmerlynn** [comment]
Just a minor addendum: please make sure you also configure fluentd-gcp.
I've actually been meaning to open an issue on the fact that the event
stream isn't actually in our cluster logging stream, and this issue happily
just solves it. We'll finally get events in the e2e GCP logs...

**Comment by satnam6502** [comment]
@saad-ali happy to help with the logging aspect of this.
Note that cluster logged can be configured to target Elasticsearch or Cloud Logging or neither.
We don't record the cluster logging choice in any systematic manner -- it's typically just an env var set in `cluster/some-provicer/config-default.sh` We should be able to query characteristics of a cluster, like "what kind of cluster logging have you been set up with?" or "what  is your mechanism for forwarding events?" and then we could build tools that work with either logging choice without imposing a lot of complexity on the client code.

**Comment by davidopp** [comment]
Also we'll need to consider how this will impact using events as an inter-component communication mechanism (see https://github.com/GoogleCloudPlatform/kubernetes/pull/3247#issuecomment-69877511).

**Comment by saad-ali** [comment]
Had a meeting with @davidopp @dchen1107 @lavalamp 

Conclusion was that we need data first and foremost. I will start working on benchmarking our etcd usage.

Once we have numbers, we can see if events actually will be a problem at the scale we intend to support for 1.0, if so we can reduce the TTL of events to something small and set up a mechanism to pull old events in to elastic search.

**Comment by dchen1107** [comment]
@saad-ali you can patch #4442 to your branch, so that cadvisor and heapster can collects etcd's resource stats for you.

**Comment by saad-ali** [comment]
Here is the data that I've gathered:
* 4 minions no extra pods other then those created with normal cluster start up:
  * [Chart](https://docs.google.com/a/google.com/spreadsheets/d/1zxTXdY1gfiIzxpbk7Ys5f8l_4Kj6FjPTzOZvRa3NLpU/pubchart?oid=656058140&format=interactive) showing memory usage with events
  * [Chart](https://docs.google.com/a/google.com/spreadsheets/d/1zxTXdY1gfiIzxpbk7Ys5f8l_4Kj6FjPTzOZvRa3NLpU/pubchart?oid=228415774&format=interactive) showing memory usage without events
    * Instead of writing events to etcd events are swallowed by the API server
* 22 minions no extra pods other then those created with normal cluster start up
  * [Chart](https://docs.google.com/a/google.com/spreadsheets/d/1zxTXdY1gfiIzxpbk7Ys5f8l_4Kj6FjPTzOZvRa3NLpU/pubchart?oid=285524783&format=interactive) showing memory usage with events
  * [Chart](https://docs.google.com/a/google.com/spreadsheets/d/1zxTXdY1gfiIzxpbk7Ys5f8l_4Kj6FjPTzOZvRa3NLpU/pubchart?oid=18661358&format=interactive) showing memory usage without events
    * Instead of writing events to etcd events are swallowed by the API server
* 4 minions generating tons of events
  * [Chart](https://docs.google.com/a/google.com/spreadsheets/d/1zxTXdY1gfiIzxpbk7Ys5f8l_4Kj6FjPTzOZvRa3NLpU/pubchart?oid=1030217379&format=interactive) showing memory usage correlated with number of events
    * Even at 30,000 events the cluster was responsive; albeit with high latency: e.g. ```kubectl get events``` took 12.4s with 5,256 events, and 74.5s with 31,400 events.

Raw data: [link](https://docs.google.com/a/google.com/spreadsheets/d/1zxTXdY1gfiIzxpbk7Ys5f8l_4Kj6FjPTzOZvRa3NLpU/edit?usp=sharing).

Observations from these tests:
* Etcd memory usage grows over the course of about 24 hours before stabilizing.
* Events don't appear to add a huge memory overhead during normal operation.
* As expected etcd's memory footprint grows linearly with number of events.

Next steps:
* Limit number of events in etcd
  * Obviously events growing unbounded in etcd is a problem.
  * Proposal
    1. Reduce the TTL of events from 2 days to 1 hour.
      * Quick and easy way to bound event growth while allowing kubectl get events, etc. to continue working the same way.
    2. Create a new process inside a container that uses the API server to fetch events and stores them to a log on disk.
      * So users who are interested in older events still have access to them.
      * fluentd/elasticsearch can be used to process the logs
    3. Create a separate instance of etcd strictly for events.
      * Because events are not strictly required for the normal function/control of the cluster, I would like to isolate them altogether.
      * This can get fancy and be a new "events server" that initially writes to a separate etcd DB, and later can be swapped out for something else.
* After minimizing/removing the events pressure from etcd, I'd like to continue a more rigorous evaluation of etcd to figure out if it will be stable at the scales we're planning for v1.0 (100 nodes and 30-50 pods/node).

**Comment by satnam6502** [comment]
Happy to help with the Fluentd -> Elasticsearch option for the events log
-- I think this should be straight forward. We could then re-wire kubectl
etc. to get the events from Elasticsearch. Note that it is possible to
create a cluster with (a) no cluster logging; (b) Fluentd -> Elasticsearch;
(c) Fluentd -> Cloud Logging. We don't have a systematic way of knowing
what the cluster level logging options are (currently specified in
/cluster/$provider/config-default.sh). Somehow we need to systematically
recored "cluster characteristics" so they can be used by tools e.g. to
query the appropriate cluster level logging system.

Satnam

**Comment by lavalamp** [comment]
> We could then re-wire kubectl etc. to get the events from Elasticsearch.

I think this needs to be optional/in addition to the ordinary way from apiserver.

**Comment by a-robinson** [comment]
+1 to @lavalamp -- we can't make the core components depend on an optional cluster add-on.

The place where elasticsearch would make more sense for this problem is as an eventual replacement for the second etcd instance that Saad proposes, where it would be a standard part of the cluster setup.

**Comment by dchen1107** [comment]
I am fine with either proposal b) or proposal c). 

According to proposal b) with fluentd->ElasticSearch, there are three different types of cluster setups. When setup option a) is chosen, the users make the decision on no historical stats; kubectl just returns the recent events related to objects. But if setup option b) and c), kubectl should provide an access point to user to retrieve such logs. We might have a proxy through apiserver? 

According to proposal c) with separate instance of etcd server, we might change it with a different infrastore / database later. 

cc/ @bgrant0607

**Comment by bgrant0607** [comment]
Optional is fine, but we can't leave users without a way to figure out what's going on. Providing a convenient access mechanism via kubectl is desirable.

**Comment by saad-ali** [comment]
To be clear, I'm planning on doing all three parts of the proposal (a, b, and c). @satnam6502, perfect--I'll sync up with you when working on b.

**Comment by vishh** [comment]
@satnam6502: Events can be stored in backends like Datadog and Influxdb in addition to Elasticsearch. Should we come up with a pluggable backend option instead? It could be etcd or elasticsearch or whatever. 
Heapster is currently well suited for pluggable backends, but it does not handle events as of now.

**Comment by davidopp** [comment]
Do we know yet whether etcd can even handle the non-event load for large clusters? If we think we eventually will need to replace etcd with something else even for non-events, then it might make more sense to keep events in the same store, but just use something more performant than etcd.

**Comment by bgrant0607** [comment]
@davidopp Think of this as infrastore. Even if we had a store that could handle the load, the events would be stored for a limited duration.

**Comment by derekwaynecarr** [comment]
Is it possible to make the charts available as public information?  I am getting permission denied when I attempt to view them from my redhat domain.

**Comment by lavalamp** [comment]
@saad-ali can you share the charts with kubernetes-dev so @derekwaynecarr and others can see them?

@davidopp what Brian said is true, we should do this in any case. However I can answer in the negative right now. I think it's contention over updating the boundPods that is causing the problem. Need to fix that and try #5274 again.

**Comment by saad-ali** [comment]
@derekwaynecarr I replaced the links in the original post to point to a doc that is shared with kubernetes-dev@googlegroups.com. Let me know if you still have trouble.

**Comment by satnam6502** [comment]
@vishn : pluggable back ends sounds like a good idea.

**Comment by saad-ali** [comment]
After sync'ing with @vishh, we agreed that we have the same long term goals (an event proxy able to pipe events to any persistent store), and so we will align our efforts (and avoid duplicating work). Instead of creating a separate process inside a container that uses the API server to fetch events and stores them to a log on disk (part b of the original proposal), I will use heapster to act as the proxy. I should be able to reuse most of the code that I've written so far. This change also aligns with heapster becoming the primary mechanism for Kubernetes users to get cluster metrics and status.

**Comment by jimmidyson** [comment]
If we're going to extend Heapster to do this, I think it makes sense to be able to run heapster in different modes: metric collection, event collection, etc. Default to all-in-one for ease of use, but separating out would probably help to make this more scalable.

**Comment by vishh** [comment]
The long term goal is to shard heapster. Do you think it is necessary to
split events and metrics even with sharding in place?

**Comment by jimmidyson** [comment]
Not that should make it unnecessary. I'm trying to think of any use cases for scaling metrics collectors & event collectors separately but I can't...

**Comment by bgrant0607** [comment]
cc @jiangyaoguo

**Comment by jiangyaoguo** [comment]
@bgrant0607 #14823 has made a a separate etcd instance to store events. And events in etcd have a  1 hour TTL, and can be collected and pushed to other persistent store by heapster. I don't  know the performance of etcd when we actually only create/read/update events. Why don't we save events to InfluxDB directly? I like the idea of @saad-ali new "events server" . Does this means a separate apiserver which only handles events? 
For now,we need a mechanism for kubectl to implement some functions like BigQuery in historical events. I‘d like to take the future work.

**Comment by quinton-hoole** [comment]
FYI, etcd performance is documented here:

https://coreos.com/etcd/docs/2.2.0/benchmarks/etcd-2-2-0-rc-benchmarks.html

In summary:

* write performance approximately 15-75 QPS per client, decreasing with the number of clients (e.g. 75 QPS for a single client, 15 QPS per client for 256 concurrent clients).
* read performance approximately 19,000 QPS per etcd server total (i.e. ~19K x 3 = 57K QPS for a 3-member etcd quorum).

The above results are for GCE n1-highcpu-2 VM's as etcd nodes on local SSD's.  Results on remote disks (as per default Kubernetes configuration) are considerably lower.

**Comment by bgrant0607** [comment]
Have we measured etcd on tmpfs?

@jiangyaoguo In order to write to another storage backend we need to abstract our internal storage interface and rebuild features provided by etcd, such as watch (which we are rebuilding over an in-memory cache) and optimistic concurrency.

cc @davidopp

**Comment by quinton-hoole** [comment]
@bgrant0607 I tried etcd on tmpfs and got roughly the same results as above (which was on local SSD).

**Comment by quinton-hoole** [comment]
@bgrant0607 PS: You'll note that the numbers above don't correlate with being I/O bound against the SSD, so my above tmpfs results make intuitive sense.

**Comment by xiang90** [comment]
@quinton-hoole It is CPU bounded on core two machines with SSD I believe.

**Comment by quinton-hoole** [comment]
@xiang90 Yes, that makes sense, and agrees with my observations.

**Comment by davidopp** [comment]
@hongchaodeng

**Comment by davidopp** [comment]
ref/ #19637

**Comment by lavalamp** [comment]
Kicking this out of 1.2 unless @davidopp reallllly wants it.

**Comment by timothysc** [comment]
xref: https://github.com/kubernetes/kubernetes/issues/41907

**Comment by krmayankk** [comment]
Has this moved at all ?

**Comment by lavalamp** [comment]
Events are in a separate etcd cluster in most set-ups. We (API Machinery) have no plans to work on this further at the moment. This would be a huge task.

Anyone wanting to revisit this should probably start with a KEP. (Actually, I'd start in an API Machinery SIG meeting before going to that much effort.) Let me repeat that this would be a huge task and it's hard to imagine a justification that would make it worth the effort.
